engine/runtime-parser/src/main/java/org/enso/compiler/core/Persistance.java

```diff
@@ -0,0 +1,534 @@
+package org.enso.compiler.core;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataInput;
+import java.io.DataInputStream;
+import java.io.DataOutput;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.ServiceLoader;
+
+import org.enso.compiler.core.Persistance.Reference;
+
+public abstract class Persistance<T> {
+  private final Class<T> clazz;
+  private final boolean includingSubclasses;
+  private final int id;
+
+  protected Persistance(Class<T> clazz, boolean includingSubclasses, int id) {
+    this.clazz = clazz;
+    this.includingSubclasses = includingSubclasses;
+    this.id = id;
+  }
+
+  protected abstract void writeObject(T obj, Output out) throws IOException;
+  protected abstract T readObject(Input in) throws IOException, ClassNotFoundException;
+
+  @Override
+  public String toString() {
+    StringBuilder sb = new StringBuilder();
+    sb.append("Persistance{");
+    sb.append("clazz=").append(clazz.getName());
+    sb.append(", id=").append(id);
+    sb.append('}');
+    return sb.toString();
+  }
+
+  public static interface Output extends DataOutput {
+    public abstract <T> void writeInline(Class<T> clazz, T obj) throws IOException;
+    public abstract void writeObject(Object obj) throws IOException;
+  }
+
+  public static interface Input extends DataInput {
+    public abstract <T> T readInline(Class<T> clazz) throws IOException;
+    public abstract Object readObject() throws IOException;
+    public <T> Reference<T> readReference(Class<T> clazz) throws IOException;
+
+    /**
+    * Looks an associated service up. The service can be specified as
+    * additional arguments to {@link #readObject(byte[], Object...)}.
+    *
+    * @param <T> type of the service requested
+    * @param clazz the class representing the requested type
+    * @return instance of the requested service or {@code null}
+    */
+    public abstract <T> T lookup(Class<T> clazz);
+  }
+
+  final void writeInline(Object obj, Output out) throws IOException {
+    writeObject(clazz.cast(obj), out);
+  }
+
+  final T readWith(Input in) {
+    try {
+      return readObject(in);
+    } catch (IOException | ClassNotFoundException ex) {
+      throw raise(RuntimeException.class, ex);
+    }
+  }
+
+  public static <T> Reference<T> readObject(byte[] arr, Object... context) throws IOException {
+    for (var i = 0; i < Generator.HEADER.length; i++) {
+      if (arr[i] != Generator.HEADER[i]) {
+        throw new IOException("Wrong header");
+      }
+    }
+    var buf = ByteBuffer.wrap(arr);
+    var at = buf.getInt(4);
+    var cache = new InputCache(buf, context);
+
+    return Reference.from(cache, at);
```

Comment:
#### Header

These are two helper methods to create a `byte[]` from a single object and then read it back. The idea is:
- 4 bytes header
- `int` describing the version
- 4 bytes to locate the beginning of the object (the objects aren't written linearly)

E.g. 12 bytes overhead before the actual data start.

#### Versioning

- when you change something really core in the `Persitance` itself - **change the header**
- when you add or remove a `Persistance` implementation the **version changes** (computed from all the IDs)
- when you change format of some `Persitance.writeObject` method - **change its ID**

Reply:
This really needs to go into comments section. No one is going to remember those guidelines or look them up in PR

Reply:
Done in 54a0471